devito/symbolics/printer.py

```diff
@@ -318,14 +356,66 @@ def _print_Fallback(self, expr):
 
 # Lifted from SymPy so that we go through our own `_print_math_func`
 for k in ('exp log sin cos tan ceiling floor').split():
-    setattr(CodePrinter, '_print_%s' % k, CodePrinter._print_math_func)
+    setattr(_DevitoPrinterBase, '_print_%s' % k, _DevitoPrinterBase._print_math_func)
 
 
 # Always parenthesize IntDiv and InlineIf within expressions
 PRECEDENCE_VALUES['IntDiv'] = 1
 PRECEDENCE_VALUES['InlineIf'] = 1
 
 
+# Sympy 1.11 has introduced a bug in `_print_Add`, so we enforce here
+# to always use the correct one from our printer
+if Version(sympy.__version__) >= Version("1.11"):
+    setattr(sympy.printing.str.StrPrinter, '_print_Add', _DevitoPrinterBase._print_Add)
+
+
+class CDevitoPrinter(_DevitoPrinterBase, C99CodePrinter):
+
+    _default_settings = {**_DevitoPrinterBase._default_settings,
+                         **C99CodePrinter._default_settings}
+    _func_litterals = {np.float32: 'f', np.complex64: 'f'}
+    _func_prefix = {np.float32: 'f', np.float64: 'f',
+                    np.complex64: 'c', np.complex128: 'c'}
+
+    # These cannot go through _print_xxx because they are classes not
+    # instances
+    type_mappings = {**C99CodePrinter.type_mappings,
+                     c_complex: 'float _Complex',
+                     c_double_complex: 'double _Complex'}
+
+    def _print_ImaginaryUnit(self, expr):
+        return '_Complex_I'
+
+
+class CXXDevitoPrinter(_DevitoPrinterBase, CXX11CodePrinter):
+
+    _default_settings = {**_DevitoPrinterBase._default_settings,
+                         **CXX11CodePrinter._default_settings}
+    _ns = "std::"
+    _func_litterals = {}
+    _func_prefix = {np.float32: 'f', np.float64: 'f'}
+
+    # These cannot go through _print_xxx because they are classes not
+    # instances
+    type_mappings = {**CXX11CodePrinter.type_mappings,
+                     c_complex: 'std::complex<float>',
+                     c_double_complex: 'std::complex<double>'}
+
+    def _print_ImaginaryUnit(self, expr):
+        return f'1i{self.prec_literal(expr).lower()}'
+
+
+class AccDevitoPrinter(CXXDevitoPrinter):
+
+    pass
+
+
+printer_registry: dict[str, type[_DevitoPrinterBase]] = {
+    'C': CDevitoPrinter, 'openmp': CDevitoPrinter,
+    'openacc': AccDevitoPrinter}
```

Comment:
it's kinda weird to see OpenACC here inside `devito/symbolics`. I'm now thinking maybe the language-specific Printers should be moved inside their respective files in `passes/iet/languages/...`

Reply:
That's what I originally had, no strong preference, but with how small the Pro part is it's a bit simpler to have it all in one place. But happy to move it back